sphinx/domains/c/_symbol.py

```diff
@@ -101,18 +101,43 @@ def __init__(
         self.isRedeclaration = False
         self._assert_invariants()
 
-        # Remember to modify Symbol.remove if modifications to the parent change.
-        self._children: list[Symbol] = []
-        self._anonChildren: list[Symbol] = []
-        # note: _children includes _anonChildren
+        # These properties store the same children for different access
+        # patterns. Symbol._add_child and Symbol._remove_child should be
+        # used for modifying them.
+        self._childrenByName: dict[str, Symbol] = {}
+        self._childrenByDocname: dict[str, dict[str, Symbol]] = {}
+        self._anonChildren: set[Symbol] = set()
+
         if self.parent:
-            self.parent._children.append(self)
+            self.parent._add_child(self)
         if self.declaration:
             self.declaration.symbol = self
 
         # Do symbol addition after self._children has been initialised.
         self._add_function_params()
 
+    @property
+    def _children(self) -> Sequence[Symbol]:
+        return list(self._childrenByName.values())
```

Comment:
Do you want to have a fixed sequence (i.e., is it possible that the child map is updated at the same time you are iterating over it from somewhere else?)

Reply:
Are you asking why I have used `list(...)`? I got this error from `mypy`:

`Incompatible return value type (got "dict_values[str, Symbol]", expected "Sequence[Symbol]")`

We don't want to expose the dict behind `children()` so I made it a list. Is there a better approach?

Reply:
The issue with mypy is because the type is a `Sequence` and `.values()` is an iterator view (more precisely a `ValuesView`), so you can change its typing.

However, when you use a `list`, you'll create a new object so if the are a lot of children, you can be *very* slow, e.g.

```python

import timeit
from random import random, randint

def y():
	return [random() for _ in range(randint(1000))]

x = {i: y() for i in range(1000)}

timeit.timeit('list(x.values())', globals={'x': x})
# 2.46

timeit.timeit('x.values()', globals={'x': x})
# 0.026
```

> We don't want to expose the dict behind the _children alias so I made it a list. Is there a better approach?

Are you talking about the possibility of writing `_children.mapping` ?

Reply:
Okay, I'll change the type to `ValuesView[Symbol]`.